api/enki.py

```diff
@@ -0,0 +1,521 @@
+                'library_id',
+		'username',
+		'password',
+		'url'
+        ]:
+            value = config.get(name)
+            if value:
+                value = value.encode("utf8")
+            values.append(value)
+        return values
+
+    @classmethod
+    def from_environment(cls, _db):
+	values = cls.environment_values()
+	if len([x for x in values if not x]):
+	    cls.log.info( "No Enki client configured" )
+	    return None
+        return cls(_db)
+
+    @property
+    def source(self):
+        return DataSource.lookup(self._db, DataSource.ENKI)
+
+    @property
+    def authorization_headers(self):
+        authorization = u":".join([self.username, self.password, self.library_id])
+        authorization = authorization.encode("utf_16_le")
+        authorization = base64.standard_b64encode(authorization)
+        return dict(Authorization="Basic " + authorization)
+
+    def refresh_bearer_token(self):
+        url = self.base_url + self.access_token_endpoint
+        headers = self.authorization_headers
+        response = self._make_request(
+            url, 'post', headers, allowed_response_codes=[200]
+        )
+        return self.parse_token(response.content)
+
+    def request(self, url, method='get', extra_headers={}, data=None,
+                params=None, exception_on_401=False):
+        """Make an HTTP request, acquiring/refreshing a bearer token
+        if necessary.
+        """
+        if not self.token:
+            self.token = self.refresh_bearer_token()
+
+        headers = dict(extra_headers)
+        headers['Authorization'] = "Bearer " + self.token
+        headers['Library'] = self.library_id
+        if exception_on_401:
+            disallowed_response_codes = ["401"]
+        else:
+            disallowed_response_codes = None
+            response = self._make_request(
+            url=url, method=method, headers=headers,
+            data=data, params=params,
+            disallowed_response_codes=disallowed_response_codes
+        )
+        if response.status_code == 401:
+            # This must be our first 401, since our second 401 will
+            # make _make_request raise a RemoteIntegrationException.
+            #
+            # The token has expired. Get a new token and try again.
+            self.token = None
+            return self.request(
+                url=url, method=method, extra_headers=extra_headers,
+                data=data, params=params, exception_on_401=True
+            )
+        else:
+            return response
+
+    def availability(self, patron_id=None, since=None, title_ids=[], strt=0, qty=2000):
+        self.log.debug ("requesting : "+ str(qty) + " books starting at econtentRecord" +  str(strt))
+        url = str(self.base_url) + str(self.availability_endpoint)
+        args = dict()
+	args['method'] = "getAllTitles"
+	args['id'] = "secontent"
+        args['strt'] = strt
+        args['qty'] = qty
+        args['lib'] = self.lib
```

Comment:
Again, it's not super important that this be tested, but in this situation I find it's useful to create a helper method that creates the data structures (URL and args, in this case). Have the method that makes the request call the helper method, test the helper method and leave the request method alone.

This lets you encode information about how the Enki API works, so that you'll find out immediately if another argument is ever introduced in the code.